mosaic/mosaic.go

```diff
@@ -0,0 +1,602 @@
+
+	// Scale the image.
+	scaledImg := m.scaleImage(img, newWidth, newHeight)
+
+	return scaledImg
+}
+
+// fitImage scales image while preserving aspect ratio.
+func (m *Mosaic) scaleImageWithoutDistortion(img image.Image, maxWidth, maxHeight int) image.Image {
+	bounds := img.Bounds()
+	srcWidth := bounds.Max.X - bounds.Min.X
+	srcHeight := bounds.Max.Y - bounds.Min.Y
+
+	// Calculate scaling factor to fit within maxWidth/maxHeight.
+	widthRatio := float64(maxWidth) / float64(srcWidth)
+	heightRatio := float64(maxHeight) / float64(srcHeight)
+
+	// Use the smaller ratio to ensure image fits.
+	ratio := math.Min(widthRatio, heightRatio)
+
+	newWidth := int(float64(srcWidth) * ratio)
+	newHeight := int(float64(srcHeight) * ratio)
+
+	// Scale the image.
+	scaledImg := m.scaleImage(img, newWidth, newHeight)
+
+	return scaledImg
+}
+
+// centerImage centers the original image without scaling.
+func (m *Mosaic) centerImage(img image.Image, maxWidth, maxHeight int) image.Image {
+	bounds := img.Bounds()
+	srcWidth := bounds.Max.X - bounds.Min.X
+	srcHeight := bounds.Max.Y - bounds.Min.Y
+
+	// If image is larger than max dimensions, scale it down.
+	if srcWidth > maxWidth || srcHeight > maxHeight {
+		return m.fitImage(img, maxWidth, maxHeight)
+	}
+
+	// Otherwise center it without scaling.
+	return m.centerScaledImage(img, maxWidth, maxHeight)
+}
+
+// centerScaledImage places a scaled image in the center of a larger canvas.
+func (m *Mosaic) centerScaledImage(img image.Image, maxWidth, maxHeight int) image.Image {
+	bounds := img.Bounds()
+	srcWidth := bounds.Max.X - bounds.Min.X
+	srcHeight := bounds.Max.Y - bounds.Min.Y
+
+	// Create a new black canvas.
+	result := image.NewRGBA(image.Rect(0, 0, maxWidth, maxHeight))
+
+	// Calculate offsets to center the image.
+	xOffset := (maxWidth - srcWidth) / 2
+	yOffset := (maxHeight - srcHeight) / 2
+
+	// Copy the image to the center of the canvas.
+	for y := 0; y < srcHeight; y++ {
+		for x := 0; x < srcWidth; x++ {
+			if x+xOffset >= 0 && x+xOffset < maxWidth && y+yOffset >= 0 && y+yOffset < maxHeight {
+				result.Set(x+xOffset, y+yOffset, img.At(x+bounds.Min.X, y+bounds.Min.Y))
+			}
+		}
+	}
+
+	return result
+}
+
+// applyDithering applies Floyd-Steinberg dithering.
+func (m *Mosaic) applyDithering(img image.Image) image.Image {
+	bounds := img.Bounds()
+	width := bounds.Max.X - bounds.Min.X
+	height := bounds.Max.Y - bounds.Min.Y
+
+	// Create a copy of the image.
+	result := image.NewRGBA(bounds)
+	for y := 0; y < height; y++ {
+		for x := 0; x < width; x++ {
+			result.Set(x, y, img.At(x+bounds.Min.X, y+bounds.Min.Y))
+		}
+	}
+
+	// Apply dithering based on color mode.
+	var levels int
+	switch m.colorMode {
+	case 0, 1: // No color or 8 colors.
+		levels = 2 // Binary dithering.
+	case 2: // 256 colors.
+		levels = 6 // 6 levels per channel (color cube).
+	case 3: // True color.
+		levels = 32 // Reduced levels for dithering.
+	}
+
+	// Floyd-Steinberg dithering.
```

Comment:
Why don't we use `image.FloydSteinberg` from the standard library?

Reply:
Oh nice, I didn't know that!